docs/dev/05-plugins.md

```diff
@@ -1,61 +1,126 @@
-[karma-closure]: https://github.com/karma-runner/karma-closure
-[karma-intellij]: https://github.com/karma-runner/karma-intellij
-[node-di]: https://github.com/vojtajina/node-di
-[karma-material-reporter]: https://github.com/ameerthehacker/karma-material-reporter
+```js
+// A factory function for our plugin, it will be called, when Karma needs to
+// instantiate a plugin. Normally it should return an instance of the service
+// conforming to the API requirements of the plugin type (more on that below),
+// but for our simple example we don't need any service and just print 
+// a message when function is called.
+function helloFrameworkFactory() {
+  console.log('Hello, world!')
+}
+
+module.exports = (config) => {
+  config.set({
+    plugins: [
+      {
+        // Declare the plugin, so Karma knows that it exists.
+        // 'factory' tells Karma that it should call `helloFrameworkFactory`
+        // function and use whatever it returns as a service for the DI token
+        // `framework:hello`.
+        'framework:hello': ['factory', helloFrameworkFactory]
+      }
+    ],
+    // Activate our plugin by specifying its unique name in the 
+    // corresponding configuration key.
+    frameworks: ['hello']
+  })
+}
+```
+
+## Injecting dependencies
+
+In "Dependency injection" section we discussed that it is possible to inject any Karma services into a plugin and interact with them. This can be done by setting an `$inject` property on the plugin's factory function to an array of DI tokens plugin wishes to interact with. Karma will pick up this property and pass requested services to the factory functions as parameters.
+
+Let's make the `hello` framework a bit more useful and make it add `hello.js` file to the `files` array. This way users of the plugin can, for example, access a function defined in `hello.js` from their tests.
+
+```js
+// Add parameters to the function to receive requested services.
+function helloFrameworkFactory(config) {
+  config.files.unshift({
+    pattern: __dirname + '/hello.js',
+    included: true,
+    served: true,
+    watched: false
+  })
+}
+
+// Declare DI tokens plugin wants to inject.
+helloFrameworkFactory.$inject = ['config']
+
+// The Karma config is unchaged and is omitted for brevity. See above example
+// for the plugin usage.
+```
+
+Note: Currently, Karma uses [node-di] library as a DI implementation. The library is more powerful than what's documented above, however, the DI implementation may change in the future, so we recommend not to rely on the node-di implementation details.
+
+## Preparing a plugin for publishing
+
+The above examples used a so-called inline plugin form. It is convenient for examples or a tiny ad-hoc plugin, but when publishing a plugin for others to use, it is usually better to export an object from an NPM package to hide extra complexity from the users of the plugins. This also gives a possibility to register multiple plugins by importing a single NPM package.
```

Comment:
I understand that you are trying to give a simple introduction by using the compact inline form, but it means later you need to issue a correction.  I think using the prefered form first with more explanation and no backtracking is better.  In particular I don't like the mixing of consumer and plugin code.

Reply:
I've updated examples to use a separate file for the plugin code. I still demonstrate the usage of the plugin from the local file, so readers can copy-paste code and run without dealing with multiple NPM packages. Also removed the section on *publishing* as it is no longer needed. Hope this is enough.